File: R/roth_vs_trad.R

retireTable <- function(ror = 1:20, retireTax = 35:5)
{
  rtaxRow <- function(r)
  {
    tr(
      if(r == retireTax[1]) th(id = "retiretax-th", div("Retirement Tax Rate"),
                               rowspan = length(retireTax)),
      th(r), map(ror, ~ td(0, id = paste0("cell-", r, "-", .x)))
    )
  }
  tags$table(
    id = "retireTable", class="width-12",
    thead(
      tr(th(), th(), th("Rate of Return on Retirement Savings", colspan=length(ror))),
      tr(th(), th(), map(ror, th))
    ),
    tbody(map(retireTax, rtaxRow))
  )
}


html(
  class = "finance",
  HTMLhead(
    titl = "Roth vs. Traditional", js = "roth_vs_trad",
    keywords = "Roth,Traditional,Retirement,Roth IRA,IRA,401(k),403(b)",
    desc = "A comparison of Roth vs. Traditional Retirement Accounts",
    date = "2018-09-15"
  ),
  body(
    navbar(),
    div(
      class="row",
      div(
        class="col width-3 inputs-panel bordered",
        h2("Input"),
        label(`for`="currAge", "Current Age"),
        input(type="number", id="currAge", value=25, step=1, min=0),
        label(`for`="retireAge", "Age at Retirement"),
        input(type="number", id="retireAge", value=65, step=1, min=0),
        label(`for`="contribs", "Yearly Contributions"),
        input(type="number", id="contribs", value=5000, min=0),
        label(`for`="currTax", "Current Tax Rate"),
        input(type="number", id="currTax", value=15, step=1, min=0, max=100),
        label(`for`="rorTaxSavings", "Rate of Return on Tax Savings"),
        input(type="number", id="rorTaxSavings", value=2, step=1, min=0),
        label(`for`="catchup", "Catch-Up Contributions?"),
        select(
          id="catchup",
          option(value=0, "None"),
          option(value=1000, "$1000 (as in an IRA)"),
          option(value=6000, "$6000 (as in a 401(k)")
        ),
        button("Calculate!", id = "submit")
      ),
      div(
        class = "col width-9 outputs-panel",
        p0(
          "This table demonstrates the value of a Roth Account over a traditional account ",
          "on the day your retire (after applicable taxes are taken out). Blue indicates that ",
          "the Roth performs better, whereas red indicates that the traditional performs better. ",
          "Hover over each cell to find the final value of each account."
        ),
        div(class="scroller-x", retireTable()),
        p0(
          "Assumptions: Tax savings from contributions to the tax-deferred account are invested ",
          "at the rate input, but the gains are taxed each year at the current tax rate. ",
          "All contributions are made at the beginning of the year and make the same ROR each year."
        )
      )
    )
  )
) %>%
  write2file(file = "docs/finance/roth_vs_trad.html")

File: js/roth_vs_trad.js
if(document.readyState === "complete")
{
  console.log("Document was ready");
  init();
} else
{
  document.addEventListener("DOMContentLoaded", init);
}

function init()
{
  document.getElementById("submit").addEventListener("click", calcMoney);
  calcMoney();
}

function roth(currAge, retireAge, contribs, ror, catchup)
{
  var out = 0;
  for(var i = currAge; i < retireAge; i++)
  {
    out = (out + contribs + (i >= 50)*catchup) * (1 + ror);
  }
  return out;
}

function trad(currAge, retireAge, contribs, ror, currTax, retireTax, rorTaxSavings, catchup)
{
  var out = 0;
  for(var i = currAge; i < retireAge; i++)
  {
    out = (out + (contribs + (i >= 50)*catchup)*currTax) * (1 + rorTaxSavings*(1 - currTax));
  }
  return out + roth(currAge, retireAge, contribs, ror, catchup)*(1 - retireTax);
}

function calcMoney()
{
  var currAge = Number(document.getElementById("currAge").value);
  var retireAge = Number(document.getElementById("retireAge").value);
  var contribs = Number(document.getElementById("contribs").value);
  var currTax = Number(document.getElementById("currTax").value) / 100;
  var rorTaxSavings = Number(document.getElementById("rorTaxSavings").value) / 100;
  var catchup = Number(document.getElementById("catchup").value);
  for(var ror = 1; ror <= 20; ror++)
  {
    for(var rtax = 5; rtax <= 20; rtax++)
    {
      var x = document.getElementById("cell-" + ror + "-" + rtax);
      var tmp = roth(currAge, retireAge, contribs, ror/100, catchup) -
        trad(currAge, retireAge, contribs, ror/100, currTax, rtax/100, rorTaxSavings, catchup);
      tmp = Math.round(tmp/1000);

      if(tmp <= -501) {x.className = "rdbu1";}
      else if(tmp <= -358) {x.className = "rdbu2";}
      else if(tmp <= -215) {x.className = "rdbu3";}
      else if(tmp <= -72) {x.className = "rdbu4";}
      else if(tmp <= 71) {x.className = "rdbu5";}
      else if(tmp <= 214) {x.className = "rdbu6";}
      else if(tmp <= 357) {x.className = "rdbu7";}
      else if(tmp <= 500) {x.className = "rdbu8";}
      else {x.className = "rdbu9"}
      x.innerHTML = "$" + tmp + "k";
    }
  }
}
